tests/framework/project/test_pipeline_discovery.py

```diff
@@ -0,0 +1,170 @@
+import sys
+import textwrap
+import warnings
+from pathlib import Path
+
+import pytest
+
+from kedro.framework.project import configure_project, find_pipelines
+
+
+@pytest.fixture
+def mock_package_name_with_pipelines(tmp_path, request):
+    package_name = "test_package"
+    pipelines_dir = tmp_path / package_name / "pipelines"
+    pipelines_dir.mkdir(parents=True)
+    (pipelines_dir / "__init__.py").touch()
+    for pipeline_name in request.param:
+        pipeline_dir = pipelines_dir / pipeline_name
+        pipeline_dir.mkdir()
+        (pipeline_dir / "__init__.py").write_text(
+            textwrap.dedent(
+                f"""
+                from kedro.pipeline import Pipeline, node, pipeline
+
+
+                def create_pipeline(**kwargs) -> Pipeline:
+                    return pipeline([node(lambda: 1, None, "{pipeline_name}")])
+                """
+            )
+        )
+    sys.path.insert(0, str(tmp_path))
+    yield package_name
+    sys.path.pop(0)
+
+    # Make sure that the `importlib_resources.files` in `find_pipelines`
+    # will point to the correct `test_package.pipelines` not from cache.
+    del sys.modules[f"{package_name}.pipelines"]
+
+
+@pytest.fixture
+def pipeline_names(request):
+    return request.param
+
+
+@pytest.mark.parametrize(
+    "mock_package_name_with_pipelines,pipeline_names",
+    [(x, x) for x in [set(), {"my_pipeline"}]],
+    indirect=True,
+)
+def test_find_pipelines(
+    mock_package_name_with_pipelines,
+    pipeline_names,
+):
+    configure_project(mock_package_name_with_pipelines)
+    pipelines = find_pipelines()
+    assert set(pipelines) == pipeline_names | {"__default__"}
+    assert sum(pipelines.values()).outputs() == pipeline_names
```

Comment:
I feel like this check isn't _quite_ strict enough since it doesn't actually test that `pipelines[pipeline_name]` maps to the pipeline with output `pipeline_name`. Maybe we should also test that `pipelines["__default__"]` is as we expect? Something like:
```
assert pipelines["__default__"].outputs() == pipeline_names
```

Reply:
> `assert pipelines["__default__"].outputs() == pipeline_names`

This is essentially what `assert sum(pipelines.values()).outputs() == pipeline_names` is testing, but `pipelines["__default__"] = sum(pipelines.values())` is explicitly written in `register_pipelines()` by default (and not in `find_pipelines`), so wouldn't be part of the unit test.